src/main/java/com/google/sps/data/PersistentUserRepository.java

```diff
@@ -24,92 +32,127 @@
     }
-    return userProfiles;
+    return userEntities;
   }
 
-  public void removeUserProfile(User user){
+  // function to fetch the user from the database
+  public Collection<User> fetchUserProfiles() {
     Query query = new Query("User");
     PreparedQuery results = datastore.prepare(query);
-    List<Entity> resultsList = results.asList(FetchOptions.Builder.withLimit(maxProfiles));
+    Collection<User> userProfiles = fetchUserEntities(results);
+    return userProfiles;
+  }
+
+  // function that sets a query filter for the results in the datastore
+  public PreparedQuery getQueryFilterForName(String name) {
+    Filter userNameFilter = new FilterPredicate("name", FilterOperator.EQUAL, name);
+    Query query = new Query("User").setFilter(userNameFilter);
+    PreparedQuery results = datastore.prepare(query);
+    return results;
+  }
 
-    String name = user.toString();
+    // function that fetches a single user, collection of users, filter on that user for name?
+  public Collection<User> fetchUsersWithName(String userName) {
+    PreparedQuery results = getQueryFilterForName(userName);
+    Collection<User> userProfiles = fetchUserEntities(results);
 
-    for (Entity entity : resultsList) {
-        String currentName = (String) entity.getProperty("name");
-        if(name.contains(currentName)) {
-            Key current = entity.getKey();
-            datastore.delete(current);
-        }
-    }
+    return userProfiles;
   }
+
   // function that adds user to the database if it does not exist already
   public void addUser(User user) {
-    allUsers = fetchUserProfiles();
+    String userName = user.getName();
+    Collection<User> userProfiles = fetchUsersWithName(userName);
     /*
-     * if the user exists in the database then nothing happens,
+     * if the userProfiles is larger than 1, that means the 
      * if it doesnt exist then it gets added
     */
-    if(!allUsers.contains(user)) {
-        allUsers.add(user);
+    if(userProfiles.size() == 0) {
         addUserToDatabase(user);
     }
-
   }
 
+  public void removeUserProfile(User user) throws Exception{
+    String inputUserName = user.getName();
+    System.out.println(inputUserName);
+    PreparedQuery results = getQueryFilterForName(inputUserName);
+    int size = results.countEntities();
+    System.out.println(size);
+
+    /*
+     *  if size is greater than 0 then there are entities available,
+     *  if size is 0 then users do not equal, so an exception is thrown
+    */
+    if(size > 0){
+        for (Entity entity : results.asIterable()) {
+            String currentName = (String) entity.getProperty("name");
+            System.out.println(currentName);
+            if(inputUserName.equals(currentName)) {
+                Key current = entity.getKey();
+                datastore.delete(current);
+            }
+
+        }
+    } else {
+        throw new Exception("Can't remove " + inputUserName + " that does not exist");
```

Comment:
supernit (L145): maybe substitute "does not exist" with "not found in datastore" (that way it's clear from the message that this is a database error, rather than some other internal inconsistency).

Reply:
Oh ok. Thanks.

Reply:
Done